tests/azure_iot_hub_client_ut.c

```diff
@@ -381,6 +474,690 @@ static void testAzureIoTHubClient_SendTelemetry_Success( void ** state )
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Unsubscribe, eAzureIoTMQTTSendFailed );
+    assert_int_equal( AzureIoTHubClient_UnsubscribeDeviceTwin( &xTestIoTHubClient ),
+                      eAzureIoTHubClientUnsubscribeFailed );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_DeviceTwinUnsubscribe_Success( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Unsubscribe, eAzureIoTMQTTSuccess );
+    assert_int_equal( AzureIoTHubClient_UnsubscribeDeviceTwin( &xTestIoTHubClient ),
+                      eAzureIoTHubClientSuccess );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_SendMethodResponse_InvalidArgFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+    uint8_t req_id = 1;
+    AzureIoTHubClientMethodRequest_t req =
+    {
+        .pucRequestID      = &req_id,
+        .usRequestIDLength = sizeof( req_id )
+    };
+
+    ( void ) state;
+
+    /* Fail SendMethodResponse when client is NULL */
+    assert_int_equal( AzureIoTHubClient_SendMethodResponse( NULL,
+                                                            &req, 200, ucTestMethodResponsePayload,
+                                                            sizeof( ucTestMethodResponsePayload ) ),
+                      eAzureIoTHubClientInvalidArgument );
+
+    /* Fail SendMethodResponse when request context is NULL */
+    assert_int_equal( AzureIoTHubClient_SendMethodResponse( &xTestIoTHubClient,
+                                                            NULL, 200, ucTestMethodResponsePayload,
+                                                            sizeof( ucTestMethodResponsePayload ) ),
+                      eAzureIoTHubClientInvalidArgument );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_SendMethodResponse_SendFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+    uint8_t req_id = 1;
+    AzureIoTHubClientMethodRequest_t req =
+    {
+        .pucRequestID      = &req_id,
+        .usRequestIDLength = sizeof( req_id )
+    };
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Publish, eAzureIoTMQTTSendFailed );
+    assert_int_equal( AzureIoTHubClient_SendMethodResponse( &xTestIoTHubClient,
+                                                            &req, 200, ucTestMethodResponsePayload,
+                                                            sizeof( ucTestMethodResponsePayload ) ),
+                      eAzureIoTHubClientPublishFailed );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_SendMethodResponse_EmptyResponseSuccess( void ** state )
```

Comment:
do we care about checking if methods are subscribed to before sending a response?

Reply:
As method response is fire and forget, do we care? We are not waiting for any response.

Reply:
I mean if someone tries to send a response for some weird reason without having first received a method call, is that something we care about? Or do we let it go through and maybe disconnect from the service (if that's what hub would do if it received a response for a method it never invoked).